pubsub/rocketmq/rocketmq.go

```diff
@@ -140,207 +265,199 @@ func (r *rocketMQ) Features() []pubsub.Feature {
-				return ErrRocketmqPublishMsg
-			}
-			r.logger.Debugf("rocketmq send result topic:%s tag:%s status:%v", req.Topic, msg.GetTags(), result.Status)
-			return nil
-		},
-		bo,
-		func(err error, d time.Duration) {
-			r.logger.Errorf("rocketmq error: fail to send message. topic:%s. Retrying...", msg.Topic)
-		},
-		func() {
-			r.logger.Infof("rocketmq successfully sent message after it previously failed. topic:%s.", msg.Topic)
-		},
-	)
+func (r *rocketMQ) getProducer() (mq.Producer, error) {
+	if r.producer != nil {
+		return r.producer, nil
+	}
+	r.producerLock.Lock()
+	defer r.producerLock.Unlock()
+	if r.producer != nil {
+		return r.producer, nil
+	}
+	producer, e := r.setUpProducer()
+	if e != nil {
+		return nil, e
+	}
+	r.producer = producer
+	return r.producer, nil
 }
 
-func newRocketMQMessage(req *pubsub.PublishRequest) *primitive.Message {
-	return primitive.NewMessage(req.Topic, req.Data).
-		WithTag(req.Metadata[metadataRocketmqTag]).
-		WithKeys([]string{req.Metadata[metadataRocketmqKey]}).
-		WithShardingKey(req.Metadata[metadataRocketmqShardingKey])
+func (r *rocketMQ) resetProducer() {
+	r.producerLock.Lock()
+	defer r.producerLock.Unlock()
+	r.producer = nil
 }
 
-type mqSubscribeCallback func(ctx context.Context, msgs ...*primitive.MessageExt) (mqc.ConsumeResult, error)
-
-func (r *rocketMQ) adaptCallback(topic, consumerGroup, mqType, mqExpr string, handler pubsub.Handler) mqSubscribeCallback {
-	return func(ctx context.Context, msgs ...*primitive.MessageExt) (mqc.ConsumeResult, error) {
-		success := true
-		for _, msg := range msgs {
-			cloudEventsMap := pubsub.NewCloudEventsEnvelope(msg.MsgId, msg.StoreHost, r.name, msg.GetProperty(primitive.PropertyKeys), msg.Topic, r.name, r.metadata.ContentType, msg.Body, "", "")
-			dataBytes, err := json.Marshal(cloudEventsMap)
-			if err != nil {
-				r.logger.Warn("rocketmq fail to marshal cloudEventsMap message, topic:%s cloudEventsMap-length:%d err:%newMessage ", msg.Topic, len(msg.Body), err)
-				success = false
-				continue
-			}
-			metadata := map[string]string{
-				metadataRocketmqType:          mqType,
-				metadataRocketmqExpression:    mqExpr,
-				metadataRocketmqConsumerGroup: consumerGroup,
-			}
-			if msg.Queue != nil {
-				metadata[metadataRocketmqBrokerName] = msg.Queue.BrokerName
-			}
-			newMessage := &pubsub.NewMessage{
-				Topic:    topic,
-				Data:     dataBytes,
-				Metadata: metadata,
-			}
-			err = handler(ctx, newMessage)
-			if err != nil {
-				r.logger.Errorf("rocketmq error: fail to process message. topic:%s cloudEventsMap-length:%d err:%v.", newMessage.Topic, len(msg.Body), err)
-				success = false
-			}
-		}
-		if !success {
-			return mqc.ConsumeRetryLater, nil
+func (r *rocketMQ) Publish(req *pubsub.PublishRequest) error {
+	r.logger.Debugf("rocketmq publish topic:%s with data:%v", req.Topic, req.Data)
+	msg := primitive.NewMessage(req.Topic, req.Data)
+	for k, v := range req.Metadata {
+		if strings.EqualFold(k, metadataRocketmqTag) {
```

Comment:
Please use a `switch strings.ToLowercase()` rather than invoking `EqualFold` every time. They are all ASCII strings so no need for EqualFOld

Reply:
ok